File: src/lib/publisher.ts
import { Post, type PostDocument } from "@/models/Post";
import {
  SocialAccount,
  type SocialAccountDocument,
} from "@/models/SocialAccount";
import { logActivity } from "@/models/ActivityLog";
import { notifyN8n } from "./n8n";
import {
  composeCaption,
  publishToFacebook,
  publishToInstagram,
} from "./social";
import {
  commentOnPost,
  publishCarouselToInstagram,
  publishReelToFacebook,
  publishReelToInstagram,
  publishStoryToFacebook,
  publishStoryToInstagram,
  publishVideoToFacebook,
  type VideoPublishResult,
} from "./social-video";
import { env } from "./env";

/**
 * Ek post ne actually publish kare ane DB ma status update kare.
 *
 * Aa j function manual "Publish now", cron dispatcher, reel studio ane
 * n8n webhook — badhi jagya thi vaparay che, jethi behaviour ek j rahe.
 *
 * Post no `postType` nakki kare che ke kaya Meta API par javanu:
 *   image → /media (IG) ke /photos (FB)
 *   reel  → REELS container (IG) ke video_reels (FB)
 *   carousel / story — potpotana endpoint
 */
export async function publishPost(
  postId: string,
  options: { onProgress?: (step: string) => void } = {},
): Promise<{
  ok: boolean;
  externalPostId?: string;
  permalink?: string;
  error?: string;
}> {
  const post = await Post.findById(postId);
  if (!post) return { ok: false, error: "Post not found" };

  if (post.status === "published") {
    return {
      ok: true,
      externalPostId: post.externalPostId ?? undefined,
      permalink: post.permalink ?? undefined,
    };
  }

  const account = await SocialAccount.findById(post.account).select(
    "+accessToken",
  );
  if (!account) {
    // A draft created before any account was connected has nothing to publish
    // to. That is not a failure of the draft — leave it as a draft so it can
    // be published once an account exists.
    const reason = post.account
      ? "That social account no longer exists. Reconnect it, or pick another."
      : "This draft has no account yet. Connect Instagram or Facebook, then choose one.";
    post.error = reason;
    await post.save();
    return { ok: false, error: reason };
  }

  const accessToken = account.accessToken || env.metaDefaultAccessToken;
  if (!accessToken) {
    post.status = "failed";
    post.error = `"${account.displayName}" mate access token set nathi`;
    await post.save();
    await logActivity({
      level: "error",
      action: "post.publish",
      message: post.error,
      post: post._id,
    });
    return { ok: false, error: post.error };
  }

  post.status = "publishing";
  post.attempts = (post.attempts ?? 0) + 1;
  post.error = undefined;
  await post.save();

  const progress = (step: string) => {
    options.onProgress?.(step);
  };

  try {
    // Caption: hashtag pehla comment ma jata hoy to caption ma na naakho.
    const message = post.firstComment
      ? post.caption
      : composeCaption(post.caption, post.hashtags ?? []);

    const result = await publishByType({
      post,
      account,
      accessToken,
      message,
      onProgress: progress,
    });

    post.status = "published";
    post.publishedAt = new Date();
    post.externalPostId = result.externalPostId;
    post.permalink = result.permalink;
    post.error = undefined;

    // Pehlo comment — hashtag block. Aa fail thay to post to gayo j che,
    // etle aakhu fail nathi ganta.
    if (post.firstComment && result.externalPostId) {
      try {
        const comment = await commentOnPost({
          mediaId: result.externalPostId,
          accessToken,
          message: post.firstComment,
        });
        post.firstCommentId = comment.id;
      } catch (error) {
        await logActivity({
          level: "warning",
          action: "post.first_comment",
          message: `Pehlo comment na thayo: ${(error as Error).message}`,
          post: post._id,
        });
      }
    }

    await post.save();

    await logActivity({
      level: "success",
      action: "post.published",
      message: `${post.platform} par ${post.postType ?? "post"} publish thayu → ${account.displayName}`,
      post: post._id,
      meta: { externalPostId: result.externalPostId, permalink: result.permalink },
    });

    await notifyN8n("post.published", {
      postId: String(post._id),
      platform: post.platform,
      postType: post.postType,
      account: account.displayName,
      externalPostId: result.externalPostId,
      permalink: result.permalink,
    });

    return {
      ok: true,
      externalPostId: result.externalPostId,
      permalink: result.permalink,
    };
  } catch (error) {
    const message = (error as Error).message;
    post.status = "failed";
    post.error = message;
    await post.save();

    await logActivity({
      level: "error",
      action: "post.failed",
      message: `Publish fail (${post.platform}/${post.postType}): ${message}`,
      post: post._id,
    });

    await notifyN8n("post.failed", {
      postId: String(post._id),
      platform: post.platform,
      error: message,
    });

    return { ok: false, error: message };
  }
}

/* ------------------------------------------------------------------ *
 *  Post type dith rasto
 * ------------------------------------------------------------------ */

async function publishByType(opts: {
  post: PostDocument;
  account: SocialAccountDocument;
  accessToken: string;
  message: string;
  onProgress: (step: string) => void;
}): Promise<VideoPublishResult> {
  const { post, account, accessToken, message, onProgress } = opts;
  const type = post.postType ?? "image";
  const isInstagram = post.platform === "instagram";

  const igUserId = account.igUserId ?? "";
  const pageId = account.pageId ?? "";

  if (isInstagram && !igUserId) {
    throw new Error(
      `"${account.displayName}" nu Instagram Business account id nathi — Accounts page ma fari connect karo.`,
    );
  }
  if (!isInstagram && !pageId) {
    throw new Error(`"${account.displayName}" nu Facebook Page id nathi.`);
  }

  switch (type) {
    case "reel": {
      const videoUrl = post.mediaUrl ?? "";
      if (!videoUrl) throw new Error("The reel has no video URL");

      if (isInstagram) {
        return publishReelToInstagram({
          igUserId,
          accessToken,
          caption: message,
          videoUrl,
          coverUrl: post.thumbnailUrl ?? undefined,
          shareToFeed: true,
          onProgress,
        });
      }

      // Facebook: pehla Reels API, e na chale to sadho feed video.
      try {
        return await publishReelToFacebook({
          pageId,
          accessToken,
          description: message,
          videoUrl,
          onProgress,
        });
      } catch (error) {
        onProgress(`Facebook Reels na chalyu (${(error as Error).message}) — feed video thi try karie chie`);
        return publishVideoToFacebook({
          pageId,
          accessToken,
          description: message,
          videoUrl,
        });
      }
    }

    case "carousel": {
      const urls = (post.mediaUrls ?? []).filter(Boolean);
      if (urls.length < 2) throw new Error("A carousel needs at least 2 images");

      if (isInstagram) {
        return publishCarouselToInstagram({
          igUserId,
          accessToken,
          caption: message,
          items: urls.map((url) => ({
            url,
            type: /\.(mp4|mov|webm)(\?|$)/i.test(url) ? ("video" as const) : ("image" as const),
          })),
          onProgress,
        });
      }

      // Facebook par carousel ek j API call ma nathi — pehli image + caption
      // mukiye chie, ane baki ni image e j post ma joda vagar alag rahe che.
      return publishToFacebook({
        pageId,
        accessToken,
        message,
        imageUrl: urls[0],
      });
    }

    case "story": {
      const mediaUrl = post.mediaUrl ?? "";
      if (!mediaUrl) throw new Error("The story has no media URL");
      const isVideo = post.mediaType === "video";

      return isInstagram
        ? publishStoryToInstagram({
            igUserId,
            accessToken,
            mediaUrl,
            mediaType: isVideo ? "video" : "image",
          })
        : isVideo
          ? publishStoryToFacebook({ pageId, accessToken, videoUrl: mediaUrl })
          : Promise.reject(
              new Error("A Facebook Story needs a video — image stories are not available through the API"),
            );
    }

    default: {
      // image ke text
      if (isInstagram) {
        return publishToInstagram({
          igUserId,
          accessToken,
          caption: message,
          imageUrl: post.mediaUrl ?? "",
        });
      }
      return publishToFacebook({
        pageId,
        accessToken,
        message,
        imageUrl: post.mediaUrl || undefined,
      });
    }
  }
}

File: python/automarketing/publisher.py
"""
Ek post ne kharekhar publish karvanu.

Aa J function badhi jagya thi vaparay che — "atyare publish karo" button,
scheduler, reel studio ane automation — jethi behaviour badhe ek j rahe.

Post no `post_type` nakki kare che ke kaya Meta API par javanu.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Callable, Optional

from bson import ObjectId

from .db import accounts as accounts_collection
from .db import connect
from .db import posts as posts_collection
from .logs import log_activity
from .social.publish import (
    PublishResult,
    comment_on_post,
    publish_carousel_instagram,
    publish_image_facebook,
    publish_image_instagram,
    publish_reel_facebook,
    publish_reel_instagram,
    publish_story_facebook,
    publish_story_instagram,
    publish_video_facebook,
)

Progress = Optional[Callable[[str], None]]


def now() -> datetime:
    return datetime.now(timezone.utc)


def compose_caption(caption: str, hashtags: list[str]) -> str:
    """Caption + hashtags ne ek publishable string ma jode."""
    if not hashtags:
        return caption
    tags = " ".join(tag if tag.startswith("#") else f"#{tag}" for tag in hashtags)
    return f"{caption}\n\n{tags}"


async def publish_post(post_id: Any, *, on_progress: Progress = None) -> dict:
    """Ek post publish kare ane DB ma status update kare."""
    await connect()

    post = await posts_collection().find_one({"_id": ObjectId(str(post_id))})
    if not post:
        return {"ok": False, "error": "Post madyo nahi"}

    if post.get("status") == "published":
        return {
            "ok": True,
            "external_id": post.get("external_id"),
            "permalink": post.get("permalink"),
        }

    account = await accounts_collection().find_one({"_id": ObjectId(str(post["account_id"]))})
    if not account:
        await posts_collection().update_one(
            {"_id": post["_id"]},
            {"$set": {"status": "failed", "error": "Social account madyu nahi"}},
        )
        return {"ok": False, "error": "Social account madyu nahi"}

    token = account.get("access_token") or ""
    if not token:
        error = f"\"{account.get('display_name')}\" mate access token nathi"
        await posts_collection().update_one(
            {"_id": post["_id"]}, {"$set": {"status": "failed", "error": error}}
        )
        await log_activity(level="error", action="post.publish", message=error)
        return {"ok": False, "error": error}

    await posts_collection().update_one(
        {"_id": post["_id"]},
        {
            "$set": {"status": "publishing", "error": None, "updated_at": now()},
            "$inc": {"attempts": 1},
        },
    )

    try:
        # Hashtag pehla comment ma jata hoy to caption ma na naakho.
        message = (
            post.get("caption", "")
            if post.get("first_comment")
            else compose_caption(post.get("caption", ""), post.get("hashtags") or [])
        )

        result = await _publish_by_type(
            post=post, account=account, token=token, message=message, on_progress=on_progress
        )

        changes: dict[str, Any] = {
            "status": "published",
            "published_at": now(),
            "external_id": result.external_id,
            "permalink": result.permalink,
            "error": None,
            "updated_at": now(),
        }

        # Pehlo comment — hashtag block. Aa fail thay to post to gayo j che,
        # etle aakhu fail nathi ganta.
        if post.get("first_comment") and result.external_id:
            try:
                changes["first_comment_id"] = await comment_on_post(
                    media_id=result.external_id,
                    access_token=token,
                    message=post["first_comment"],
                )
            except Exception as error:  # noqa: BLE001
                await log_activity(
                    level="warning",
                    action="post.first_comment",
                    message=f"Pehlo comment na thayo: {error}",
                )

        await posts_collection().update_one({"_id": post["_id"]}, {"$set": changes})

        await log_activity(
            level="success",
            action="post.published",
            message=(
                f"{post.get('platform')} par {post.get('post_type', 'post')} publish "
                f"thayu → {account.get('display_name')}"
            ),
            meta={"post_id": str(post["_id"]), "permalink": result.permalink},
        )

        return {
            "ok": True,
            "external_id": result.external_id,
            "permalink": result.permalink,
        }

    except Exception as error:  # noqa: BLE001
        message = str(error)[:800]
        await posts_collection().update_one(
            {"_id": post["_id"]},
            {"$set": {"status": "failed", "error": message, "updated_at": now()}},
        )
        await log_activity(
            level="error",
            action="post.failed",
            message=f"Publish fail ({post.get('platform')}/{post.get('post_type')}): {message}",
            meta={"post_id": str(post["_id"])},
        )
        return {"ok": False, "error": message}


async def _publish_by_type(
    *, post: dict, account: dict, token: str, message: str, on_progress: Progress
) -> PublishResult:
    post_type = post.get("post_type") or "image"
    is_instagram = post.get("platform") == "instagram"

    ig_user_id = str(account.get("ig_user_id") or "")
    page_id = str(account.get("page_id") or "")

    if is_instagram and not ig_user_id:
        raise ValueError(
            f"\"{account.get('display_name')}\" nu Instagram Business account id nathi — "
            "Accounts page ma fari connect karo."
        )
    if not is_instagram and not page_id:
        raise ValueError(f"\"{account.get('display_name')}\" nu Facebook Page id nathi.")

    media_url = str(post.get("media_url") or "")

    if post_type == "reel":
        if not media_url:
            raise ValueError("Reel nu video URL nathi")

        if is_instagram:
            return await publish_reel_instagram(
                ig_user_id=ig_user_id,
                access_token=token,
                caption=message,
                video_url=media_url,
                cover_url=str(post.get("thumbnail_url") or ""),
                share_to_feed=True,
                on_progress=on_progress,
            )

        # Facebook: pehla Reels API, e na chale to sadho feed video.
        try:
            return await publish_reel_facebook(
                page_id=page_id,
                access_token=token,
                description=message,
                video_url=media_url,
                on_progress=on_progress,
            )
        except Exception as error:  # noqa: BLE001
            if on_progress:
                on_progress(f"Facebook Reels na chalyu ({error}) — feed video thi try karie chie")
            return await publish_video_facebook(
                page_id=page_id, access_token=token, description=message, video_url=media_url
            )

    if post_type == "carousel":
        urls = [u for u in (post.get("media_urls") or []) if u]
        if len(urls) < 2:
            raise ValueError("Carousel mate ochha ma ochhi 2 image joiye")

        if is_instagram:
            return await publish_carousel_instagram(
                ig_user_id=ig_user_id,
                access_token=token,
                caption=message,
                items=[
                    {
                        "url": url,
                        "type": "video" if url.lower().split("?")[0].endswith((".mp4", ".mov", ".webm")) else "image",
                    }
                    for url in urls
                ],
                on_progress=on_progress,
            )

        # Facebook par carousel ek j API call ma nathi — pehli image mukiye chie.
        return await publish_image_facebook(
            page_id=page_id, access_token=token, message=message, image_url=urls[0]
        )

    if post_type == "story":
        if not media_url:
            raise ValueError("Story nu media URL nathi")
        is_video = post.get("media_type") == "video"

        if is_instagram:
            return await publish_story_instagram(
                ig_user_id=ig_user_id,
                access_token=token,
                media_url=media_url,
                is_video=is_video,
            )
        if not is_video:
            raise ValueError("Facebook Story mate video joiye — image story API thi nathi thati")
        return await publish_story_facebook(
            page_id=page_id, access_token=token, video_url=media_url
        )

    # image ke text
    if is_instagram:
        return await publish_image_instagram(
            ig_user_id=ig_user_id,
            access_token=token,
            caption=message,
            image_url=media_url,
        )
    return await publish_image_facebook(
        page_id=page_id, access_token=token, message=message, image_url=media_url
    )
